File: 250314Shenfeng/250501rnn.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <algorithm>
#include <map>
#include <chrono>
#include <iomanip>
#include <Eigen/Dense>
#include <torch/torch.h>
#include <filesystem>

// Helper: Parse CSV (Date, Count)
struct Record {
    std::tm date;
    double count;
};

std::vector<Record> load_csv(const std::string& filename, int date_col_idx = 0, int count_col_idx = 1) {
    std::vector<Record> data;
    std::ifstream file(filename);
    std::string line;
    std::getline(file, line); // header
    while (std::getline(file, line)) {
        std::stringstream ss(line);
        std::string cell;
        std::vector<std::string> cells;
        while (std::getline(ss, cell, ',')) cells.push_back(cell);
        if (cells.size() <= std::max(date_col_idx, count_col_idx)) continue;
        std::string date_str = cells[date_col_idx];
        std::string count_str = cells[count_col_idx];
        std::tm tm = {};
        if (date_str.find(":") != std::string::npos) {
            std::istringstream ss(date_str);
            ss >> std::get_time(&tm, "%d/%m/%Y %H:%M");
        } else {
            std::istringstream ss(date_str);
            ss >> std::get_time(&tm, "%d/%m/%Y");
        }
        if (!date_str.empty() && !count_str.empty()) {
            data.push_back({tm, std::stod(count_str)});
        }
    }
    return data;
}

// Convert tm to YYYY-MM-DD string
std::string tm_to_string(const std::tm& tm) {
    char buf[16];
    std::strftime(buf, sizeof(buf), "%Y-%m-%d", &tm);
    return std::string(buf);
}

// Aggregate to daily counts
std::map<std::string, double> aggregate_daily(const std::vector<Record>& data) {
    std::map<std::string, double> daily_counts;
    for (const auto& r : data) {
        std::string day = tm_to_string(r.date);
        daily_counts[day] += 1.0;
    }
    return daily_counts;
}

// MinMaxScaler
struct MinMaxScaler {
    double min, max;
    MinMaxScaler() : min(0), max(1) {}
    void fit(const std::vector<double>& v) {
        min = *std::min_element(v.begin(), v.end());
        max = *std::max_element(v.begin(), v.end());
    }
    std::vector<double> transform(const std::vector<double>& v) {
        std::vector<double> res;
        for (auto x : v) res.push_back((x - min) / (max - min));
        return res;
    }
    std::vector<double> inverse_transform(const std::vector<double>& v) {
        std::vector<double> res;
        for (auto x : v) res.push_back(x * (max - min) + min);
        return res;
    }
};

// Create sequences for RNN
void create_sequences(const std::vector<double>& data, int time_step, std::vector<std::vector<double>>& X, std::vector<double>& y) {
    for (size_t i = 0; i + time_step < data.size(); ++i) {
        X.push_back(std::vector<double>(data.begin() + i, data.begin() + i + time_step));
        y.push_back(data[i + time_step]);
    }
}

// RNN Model
struct RNNNet : torch::nn::Module {
    torch::nn::RNN rnn;
    torch::nn::Linear fc;
    RNNNet(int input_size, int hidden_size, int output_size, int num_layers)
        : rnn(torch::nn::RNNOptions(input_size, hidden_size).num_layers(num_layers).batch_first(true)),
          fc(hidden_size, output_size) {
        register_module("rnn", rnn);
        register_module("fc", fc);
    }
    torch::Tensor forward(torch::Tensor x) {
        auto out_tuple = rnn->forward(x);
        auto out = std::get<0>(out_tuple);
        out = fc->forward(out.index({torch::indexing::Slice(), -1, torch::indexing::Slice()}));
        return out;
    }
};

int main() {
    // Dataset group (example: group1)
    std::vector<std::string> files = {
        "250314Shenfeng/datasets/dfb-fire-2023-opendata.csv",
        "250314Shenfeng/datasets/df-opendata-2016-to-2017-with-stn-area.csv",
        "250314Shenfeng/datasets/df-opendata-2020-to-2022-with-stn-area.csv",
        "250314Shenfeng/datasets/df-opendata-2018-to-2019-with-stn-area.csv"
    };
    std::vector<Record> all_data;
    for (const auto& file : files) {
        auto data = load_csv(file);
        all_data.insert(all_data.end(), data.begin(), data.end());
    }
    std::sort(all_data.begin(), all_data.end(), [](const Record& a, const Record& b) {
        return std::mktime(const_cast<std::tm*>(&a.date)) < std::mktime(const_cast<std::tm*>(&b.date));
    });
    auto daily_counts_map = aggregate_daily(all_data);
    std::vector<std::string> dates;
    std::vector<double> counts;
    for (const auto& kv : daily_counts_map) {
        dates.push_back(kv.first);
        counts.push_back(kv.second);
    }
    std::cout << "Start Date: " << dates.front() << std::endl;
    std::cout << "End Date: " << dates.back() << std::endl;
    std::cout << "Total Days: " << dates.size() << std::endl;
    // Scaling
    MinMaxScaler scaler;
    scaler.fit(counts);
    auto scaled = scaler.transform(counts);
    // Train/test split
    int train_size = static_cast<int>(scaled.size() * 0.8);
    std::vector<double> train_data(scaled.begin(), scaled.begin() + train_size);
    std::vector<double> test_data(scaled.begin() + train_size, scaled.end());
    // Create sequences
    int time_step = 10;
    std::vector<std::vector<double>> X_train, X_test;
    std::vector<double> y_train, y_test;
    create_sequences(train_data, time_step, X_train, y_train);
    create_sequences(test_data, time_step, X_test, y_test);
    // Convert to torch tensors
    auto vec2tensor = [](const std::vector<std::vector<double>>& X) {
        return torch::tensor(X).unsqueeze(-1).to(torch::kFloat32);
    };
    auto vec1tensor = [](const std::vector<double>& y) {
        return torch::tensor(y).to(torch::kFloat32);
    };
    auto X_train_tensor = vec2tensor(X_train);
    auto y_train_tensor = vec1tensor(y_train);
    auto X_test_tensor = vec2tensor(X_test);
    auto y_test_tensor = vec1tensor(y_test);
    // Model
    int input_size = 1, hidden_size = 50, output_size = 1, num_layers = 2;
    RNNNet model(input_size, hidden_size, output_size, num_layers);
    model.train();
    torch::optim::Adam optimizer(model.parameters(), torch::optim::AdamOptions(0.001));
    int epochs = 20;
    for (int epoch = 0; epoch < epochs; ++epoch) {
        auto output = model.forward(X_train_tensor);
        auto loss = torch::mse_loss(output.squeeze(), y_train_tensor);
        optimizer.zero_grad();
        loss.backward();
        optimizer.step();
        std::cout << "Epoch " << epoch+1 << ", Loss: " << loss.item<float>() << std::endl;
    }
    // Predict
    model.eval();
    auto train_pred = model.forward(X_train_tensor).detach().squeeze().to(torch::kCPU);
    auto test_pred = model.forward(X_test_tensor).detach().squeeze().to(torch::kCPU);
    // Inverse transform
    std::vector<double> train_pred_vec(train_pred.data_ptr<float>(), train_pred.data_ptr<float>() + train_pred.numel());
    std::vector<double> y_train_vec(y_train_tensor.data_ptr<float>(), y_train_tensor.data_ptr<float>() + y_train_tensor.size(0));
    std::vector<double> test_pred_vec(test_pred.data_ptr<float>(), test_pred.data_ptr<float>() + test_pred.numel());
    std::vector<double> y_test_vec(y_test_tensor.data_ptr<float>(), y_test_tensor.data_ptr<float>() + y_test_tensor.size(0));
    auto train_pred_actual = scaler.inverse_transform(train_pred_vec);
    auto y_train_actual = scaler.inverse_transform(y_train_vec);
    auto test_pred_actual = scaler.inverse_transform(test_pred_vec);
    auto y_test_actual = scaler.inverse_transform(y_test_vec);
    // Output results to CSV
    std::ofstream out("250501rnn_train_results.csv");
    out << "Actual,Predicted\n";
    for (size_t i = 0; i < train_pred_actual.size(); ++i)
        out << y_train_actual[i] << "," << train_pred_actual[i] << "\n";
    out.close();
    std::ofstream out2("250501rnn_test_results.csv");
    out2 << "Actual,Predicted\n";
    for (size_t i = 0; i < test_pred_actual.size(); ++i)
        out2 << y_test_actual[i] << "," << test_pred_actual[i] << "\n";
    out2.close();
    // Forecast next 30 days
    std::vector<double> last_sequence(scaled.end() - time_step, scaled.end());
    auto last_seq_tensor = torch::tensor(last_sequence).unsqueeze(0).unsqueeze(-1).to(torch::kFloat32);
    std::vector<double> future_predictions;
    for (int i = 0; i < 30; ++i) {
        auto next_pred = model.forward(last_seq_tensor).item<float>();
        future_predictions.push_back(next_pred);
        // Update sequence
        std::vector<double> seq(last_sequence.begin() + 1, last_sequence.end());
        seq.push_back(next_pred);
        last_seq_tensor = torch::tensor(seq).unsqueeze(0).unsqueeze(-1).to(torch::kFloat32);
        last_sequence = seq;
    }
    auto future_pred_actual = scaler.inverse_transform(future_predictions);
    // Output future predictions to CSV
    std::ofstream out3("250501rnn_future_results.csv");
    out3 << "Day,Predicted\n";
    for (size_t i = 0; i < future_pred_actual.size(); ++i)
        out3 << i+1 << "," << future_pred_actual[i] << "\n";
    out3.close();
    // Metrics
    double sse = 0, mae = 0;
    for (size_t i = 0; i < test_pred_actual.size(); ++i) {
        double err = y_test_actual[i] - test_pred_actual[i];
        sse += err * err;
        mae += std::abs(err);
    }
    double mse = sse / test_pred_actual.size();
    double rmse = std::sqrt(mse);
    mae /= test_pred_actual.size();
    std::cout << "SSE: " << sse << std::endl;
    std::cout << "RMSE: " << rmse << std::endl;
    std::cout << "MAE: " << mae << std::endl;
    // Save model
    auto t = std::time(nullptr);
    auto tm = *std::localtime(&t);
    std::ostringstream oss;
    oss << "rnn_model_" << std::put_time(&tm, "%Y%m%d_%H%M%S") << ".pth";
    torch::save(model, oss.str());
    std::cout << "Model saved as " << oss.str() << std::endl;
    return 0;
}

File: 250314Shenfeng/250501rnn.py
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error, mean_absolute_error
import torch
import torch.nn as nn
from torch.utils.data import DataLoader, TensorDataset

# ------------------ Dataset Configuration ------------------
dataset_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), ".", "datasets"))

import os


dataset_groups = {
    "group1": [
        "dfb-fire-2023-opendata.csv",
        "df-opendata-2016-to-2017-with-stn-area.csv",
        "df-opendata-2020-to-2022-with-stn-area.csv",
        "df-opendata-2018-to-2019-with-stn-area.csv"
    ],
    "group2": [
        "2013-2015-dfb-ambulance.csv"
    ],
    "group3": [
        "2023-open-data-dfb-ambulance.csv",
        "da-opendata-2018-to-2019-with-stn-area.csv",
        "da-opendata-2016-to-2017-with-stn-area.csv"
    ],
    "group4": [
        "da-opendata-2020-to-2022-with-stn-area.csv"
    ],
    "group5": [
        "2013-2015-dfb-fire.csv"
    ],
    "group6": [
        "dccfirebrigadeambulanceincidents2011.csv",
        "dccfirebrigadeambulanceincidents2012.csv"
    ]
}



# ------------------ Load and Combine Datasets ------------------

def combine_dataset_group(group_name):
    file_list = dataset_groups[group_name]
    dfs = []
    for file in file_list:
        path = os.path.join(dataset_dir, file)
        if file.endswith(".csv"):
            dfs.append(pd.read_csv(path))
        elif file.endswith(".xlsx"):
            dfs.append(pd.read_excel(path))
        else:
            raise ValueError(f"Unsupported file format: {file}")
    return pd.concat(dfs, ignore_index=True)

#selected_files = ["dfb_ambulance_2013_2015", "dfb_ambulance_2023", "da_2020_2022"]
table1 = combine_dataset_group(list(dataset_groups.keys())[0])

# ------------------ Preprocess Date and Resample ------------------
table1.columns = [col.strip().lower() for col in table1.columns]
date_col = [col for col in table1.columns if 'date' in col][0]
table1[date_col] = pd.to_datetime(table1[date_col], errors='coerce')
table1.dropna(subset=[date_col], inplace=True)
table1.set_index(date_col, inplace=True)
table1.sort_index(inplace=True)

daily_counts = table1.resample('D').size().to_frame(name='Count')
print("Start Date:", daily_counts.index.min())
print("End Date:", daily_counts.index.max())
print("Total Days:", (daily_counts.index.max() - daily_counts.index.min()).days)
print("Inferred Frequency:", pd.infer_freq(daily_counts.index))

# ------------------ Normalize and Sequence Preparation ------------------
scaler = MinMaxScaler()
scaled_data = scaler.fit_transform(daily_counts)
train_size = int(len(scaled_data) * 0.8)
train_data = scaled_data[:train_size]
test_data = scaled_data[train_size:]

def create_sequences(data, time_step=10):
    X, y = [], []
    for i in range(len(data) - time_step):
        X.append(data[i:i + time_step, 0])
        y.append(data[i + time_step, 0])
    return np.array(X), np.array(y)

time_step = 10
X_train, y_train = create_sequences(train_data, time_step)
X_test, y_test = create_sequences(test_data, time_step)

X_train = torch.tensor(X_train, dtype=torch.float32).unsqueeze(-1)
y_train = torch.tensor(y_train, dtype=torch.float32)
X_test = torch.tensor(X_test, dtype=torch.float32).unsqueeze(-1)
y_test = torch.tensor(y_test, dtype=torch.float32)

train_loader = DataLoader(TensorDataset(X_train, y_train), batch_size=32, shuffle=True)
test_loader = DataLoader(TensorDataset(X_test, y_test), batch_size=32, shuffle=False)

# ------------------ Define RNN Model ------------------
class RNNModel(nn.Module):
    def __init__(self, input_size, hidden_size, output_size, num_layers):
        super(RNNModel, self).__init__()
        self.rnn = nn.RNN(input_size, hidden_size, num_layers, batch_first=True)
        self.fc = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        out, _ = self.rnn(x)
        return self.fc(out[:, -1, :])

model = RNNModel(input_size=1, hidden_size=50, output_size=1, num_layers=2)
criterion = nn.MSELoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

# ------------------ Train the Model ------------------
for epoch in range(20):
    model.train()
    for X_batch, y_batch in train_loader:
        outputs = model(X_batch).squeeze()
        loss = criterion(outputs, y_batch)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    print(f"Epoch [{epoch+1}/20], Loss: {loss.item():.4f}")

# ------------------ Evaluate and Forecast ------------------
model.eval()
train_predictions = model(X_train).detach().numpy()
test_predictions = model(X_test).detach().numpy()

train_predictions = scaler.inverse_transform(train_predictions)
test_predictions = scaler.inverse_transform(test_predictions)
y_train = scaler.inverse_transform(y_train.numpy().reshape(-1, 1))
y_test = scaler.inverse_transform(y_test.numpy().reshape(-1, 1))

# Forecast future
last_sequence = torch.tensor(scaled_data[-time_step:], dtype=torch.float32).unsqueeze(0).unsqueeze(-1)
future_predictions = []
for _ in range(30):
    next_pred = model(last_sequence)
    next_pred = next_pred.unsqueeze(-1)
    future_predictions.append(next_pred.item())
    last_sequence = torch.cat((last_sequence[:, 1:, :], next_pred), dim=1)

future_predictions = scaler.inverse_transform(np.array(future_predictions).reshape(-1, 1))
future_dates = pd.date_range(start=daily_counts.index[-1] + pd.Timedelta(days=1), periods=30)

# ------------------ Metrics ------------------
sse = np.sum((y_test - test_predictions) ** 2)
rmse = np.sqrt(mean_squared_error(y_test, test_predictions))
mae = mean_absolute_error(y_test, test_predictions)
print(f"SSE: {sse:.2f}\nRMSE: {rmse:.2f}\nMAE: {mae:.2f}")

# ------------------ Save the Model ------------------
timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
model_filename = f"rnn_model_{timestamp}.pth"
torch.save(model.state_dict(), model_filename)
print(f"Model saved as {model_filename}")

# ------------------ Plot Results ------------------
test_dates = daily_counts.index[len(train_data) + time_step:]

plt.figure(figsize=(12, 6))
plt.plot(test_dates, y_test, label='Actual')
plt.plot(test_dates, test_predictions, label='Predicted', linestyle='--')
plt.title('Actual vs Predicted (Test Set)')
plt.legend()
plt.show()

plt.figure(figsize=(12, 6))
plt.plot(daily_counts.index, daily_counts['Count'], label='Historical')
plt.plot(future_dates, future_predictions, label='Forecast', color='red')
plt.title('Next 30 Days Forecast')
plt.legend()
plt.show()
